mypyc/irbuild/ll_builder.py

```diff
@@ -1550,10 +1552,33 @@ def check_tagged_short_int(self, val: Value, line: int, negated: bool = False) -
 
     def compare_strings(self, lhs: Value, rhs: Value, op: str, line: int) -> Value:
         """Compare two strings"""
+
+        def is_string_literal(value: Value) -> TypeGuard[LoadLiteral]:
+            return isinstance(value, LoadLiteral) and is_str_rprimitive(value.type)
+
         if op == "==":
+            if is_string_literal(lhs):
+                if is_string_literal(rhs):
+                    # we can optimize out the check entirely in some constant-folded cases
+                    return self.true() if lhs.value == rhs.value else self.false()
+                literal_length = Integer(len(lhs.value), c_pyssize_t_rprimitive, line)  # type: ignore [arg-type]
+                return self.primitive_op(str_eq_literal, [rhs, lhs, literal_length], line)
+            elif is_string_literal(rhs):
+                literal_length = Integer(len(rhs.value), c_pyssize_t_rprimitive, line)  # type: ignore [arg-type]
+                return self.primitive_op(str_eq_literal, [lhs, rhs, literal_length], line)
             return self.primitive_op(str_eq, [lhs, rhs], line)
         elif op == "!=":
-            eq = self.primitive_op(str_eq, [lhs, rhs], line)
+            if is_string_literal(lhs):
```

Comment:
looking at this again, I think we can just refactor this whole block for "!=" into:

return self.add(ComparisonOp(compare_strings(lhs, rhs, line), self.false(), ComparisonOp.EQ, line)